---
int32[] handles
int32[] zones
